### DISCUSSION TITLE
More built-in progression options

### DISCUSSION THREAD

**Discussion by yrinkel** [astashov/liftosaur#155] [Ideas]
At the very least, Liftosaur should contain the two following progression options:

[Dynamic Double Progression](https://evolvedtrainingsystems.com/overload-utilization-of-dynamic-double-progression/#:~:text=Dynamic%20Double%20Progression%20is%20a,on%20its%20own%20progression%20track.)

[Immediate Double Progression](https://legionathletics.com/double-progression/?srsltid=AfmBOordlBPmkKVo9mJT7PwDUdL_fyn_4UsWaRwWLKagxuV7tf94v6Uw)
(Which is the same as Double Progression except that the reps go down -and weight up-as soon as a set reaches the top number of reps within RPE, and the weight is immediately set down -and nb of reps back up- if the bottom nb of reps is not achieved or not within RPE.)

**Comment by snashikkar** [comment]
Yep including RPE based progression and dynamic updates to subsequent sets would be amazing to support auto regulation! 

For progress:
During either linear, double or sumreps progression,  include a flag to toggle RPE considerations on or off.
If all other criteria are met for that progression, AND iff the actual RPE is <= target RPE, then progress else do not.

For updates: 
1. if actual RPE is greater than> target RPE, then adjust weights for subsequent sets to lower weight by  using the rpe tables to determine the target weights for the subsequent sets and target RPE. So if my first set had an actual RPE of 9 vs. 7.5 as the original target RPE , then the next sets weigtts should be updated so that the user is able to complete all reps at the target RPE . 
2. If actual RPE is equal then do nothing. 
3. If actual RPE is less then target RPE, adjust weights as per scenario 1.but increase weights instead of lowering.

**Comment by petemounce** [comment]
I am embarking into http://www.startbodyweight.com/p/start-bodyweight-basic-routine.html. 

This is a routine that at its core wants you to do a given exercise variant (like pull-ups, push-ups) 3 sets of 4-8 reps, increasing 1 rep each day until 888, then progressing by changing the exercise variant to a more challenging one (for example, push-ups start as wall push-ups, then box push-ups, then quarter push-ups .... decline one arm push ups).

Each exercise variant is based on one's body weight. I think it's outside of my clue to figure out what the _effective_ weight being moved is via each exercise variant. I'm not sure the charts and so forth will therefore be so useful - but I think my motivation is likely to derive from the progression via rep count then challenging exercise variants I had not previously been able to do.

I wonder whether it might be possible to have a progression-option of "now do this other exercise variant"?

I appreciate that would make for a complex workout programme to encode that routine, though, probably.

Perhaps it might be workable with syntax like the exercise name being an array...? For example one entry from my current workout is:

```txt
Squat, Bodyweight / 3x4-8 60s / progress: custom() {~
  if (completedReps[2] >= reps[2]) {
    reps += 1
  }
~}
```

... and perhaps that could become:

```txt
["Assisted Squat, Bodyweight", "Assisted Deep Squat, Bodyweight", "Squat, Bodyweight", "Deep Squat, Bodyweight"] / 3x4-8 60s / progress: custom() {~
  if (completedReps[2] >= reps[2]) {
    reps += 1
  }
  if (completedReps[2] >= 8) {
    progressNextVariant()
  }
~}
```

?

I appreciate there's now more syntax-noise because of the need to quote. I guess `|` could be a progression separator instead of re-using `,`, maybe.

Or perhaps one might define exercise arrays as variables up-front, make it look more like:

```txt
@squat_progression = ["Assisted Squat, Bodyweight", "Assisted Deep Squat, Bodyweight", "Squat, Bodyweight", "Deep Squat, Bodyweight"]
@squat_progression / 3x4-8 60s / progress: custom() {~
  if (completedReps[2] >= reps[2]) {
    reps += 1
  }
  if (completedReps[2] >= 8) {
    progressNextVariant()
  }
~}
```

**Reply by bjornmartensson** [reply]
This seems to have some overlap with #173 (Alternative exercises in a program). You may want to add to that idea.
